### DISCUSSION TITLE
Yolov5 + React

### DISCUSSION THREAD

**Discussion by alessandrogpz** [ultralytics/yolov5#2738] [Q&A] [answered]
I would like to know if it is possible to integrate a model (.pt) created using Yolov5 into a react project. My group is preparing a landing page for a product that uses object recognition as one of the features and I would like to know how can I make the link in between the created model and react. Any help on the deployment of a model is also very welcome since I have never deployed one yet.

**Comment by glenn-jocher** [comment, marked as answer]
@alessandrogpz hi buddy. You're looking for Tensorflow.js https://www.tensorflow.org/js. 

You'd want to do a TF js export of a small model like YOLOv5s and then you can cache and run the model in the browser on the user's local resources.

The best resource for getting started with YOLOv5 TF is PR https://github.com/ultralytics/yolov5/pull/1127

**Reply by pravastacaraka** [reply]
@glenn-jocher I have a problem when exporting the model to TF.js format along with NMS. In PR #1127 the CombinedNonMaxSupression () ops is used. However, TF.js does not yet support these ops.

@alessandrogpz Do you already have a solution?

**Reply by glenn-jocher** [reply]
@pravastacaraka we're working on supporting JS export in the future, at the moment I can't be of much help, but I see a dedicated JS NMS implementation here:
https://js.tensorflow.org/api/latest/#image.nonMaxSuppression

**Reply by pravastacaraka** [reply]
@glenn-jocher in https://js.tensorflow.org/api/latest/#image.nonMaxSuppression is must 2D tensor input of boxes, but in the PR #1127 use not 2D tensor of boxes variable.

```
# Add TensorFlow NMS
if opt.tf_nms:
    boxes = tf.expand_dims(xywh2xyxy(x[0][..., :4]), 2)
    probs = x[0][:, :, 4:5]
    classes = x[0][:, :, 5:]
    scores = probs * classes
    nms = tf.image.combined_non_max_suppression(
        boxes, scores, opt.topk_per_class, opt.topk_all, opt.iou_thres, opt.score_thres, clip_boxes=False)
    return nms, x[1]
```

**Reply by glenn-jocher** [reply]
@pravastacaraka I think the opt.tf_nms `if` statement you have there is only for running TF models with detect.py using NMS. It doesn't stick any NMS modules onto exported models.

**Reply by pravastacaraka** [reply]
@glenn-jocher no,, it's exported as well. When I have exported it to the tfjs model then tried to run it there is an error that CombinedNonMaxSupression () is not supported by tfjs yet.

**Reply by glenn-jocher** [reply]
@pravastacaraka right, combined NMS is not supported so you probably want to swap it for one of the NMS functions in https://js.tensorflow.org/api/latest/#image.nonMaxSuppression that are supported.

**Reply by pravastacaraka** [reply]
@glenn-jocher Yes I want to do that but the shape of the 'boxes' variable is not the 2D tensor. So I'm still confused about how to exchange it?

**Reply by glenn-jocher** [reply]
I'm not sure, I've not worked with this particular export pipeline, but I assume that yes you'll need to cast and reshape the data to the expected input datatype and shape.

**Reply by pravastacaraka** [reply]
@glenn-jocher can you help me how to reshape the output of the boxes from this particular code?

```python
# Add TensorFlow NMS
if opt.tf_nms:
    boxes = tf.expand_dims(xywh2xyxy(x[0][..., :4]), 2)
    probs = x[0][:, :, 4:5]
    classes = x[0][:, :, 5:]
    scores = probs * classes
    nms = tf.image.combined_non_max_suppression(
        boxes, scores, opt.topk_per_class, opt.topk_all, opt.iou_thres, opt.score_thres, clip_boxes=False)
    return nms, x[1]
```

**Reply by glenn-jocher** [reply]
@pravastacaraka you can use tf.reshape to reshape the boxes to the shape you need:
https://www.tensorflow.org/api_docs/python/tf/reshape

**Comment by rudro12356** [comment]
Hi, would the steps be similar for integrating the yolov5 model into react-native app? Me and my group is developing a cross-platform app for car detection for our school's senior year project. I have been custom training the model on Colab. Any suggestions and help would be highly appreciated!
